sklearn/partial_dependence.py

```diff
@@ -0,0 +1,648 @@
+
+    Parameters
+    ----------
+    X : ndarray
+        The data
+    percentiles : tuple of floats
+        The percentiles which are used to construct the extreme
+        values of the grid axes.
+    grid_resolution : int
+        The number of equally spaced points that are placed
+        on the grid.
+
+    Returns
+    -------
+    grid : ndarray
+        All data points on the grid; ``grid.shape[1] == X.shape[1]``
+        and ``grid.shape[0] == grid_resolution * X.shape[1]``.
+    axes : seq of ndarray
+        The axes with which the grid has been created.
+    """
+    if len(percentiles) != 2:
+        raise ValueError('percentile must be tuple of len 2')
+    if not all(0. <= x <= 1. for x in percentiles):
+        raise ValueError('percentile values must be in [0, 1]')
+
+    axes = []
+    emp_percentiles = mquantiles(X, prob=percentiles, axis=0)
+    for col in range(X.shape[1]):
+        uniques = np.unique(X[:, col])
+        if uniques.shape[0] < grid_resolution:
+            # feature has low resolution use unique vals
+            axis = uniques
+        else:
+            # create axis based on percentiles and grid resolution
+            axis = np.linspace(emp_percentiles[0, col],
+                               emp_percentiles[1, col],
+                               num=grid_resolution, endpoint=True)
+        axes.append(axis)
+
+    return cartesian(axes), axes
+
+
+def _exact_partial_dependence(est, target_variables, grid, X, output=None):
+    """Calculate the partial dependence of ``target_variables``.
+
+    The function will be calculated by calling the ``predict_proba`` method of
+    ``est`` for classification or ``predict`` for regression on ``X`` for every
+    point in the grid.
+
+    Parameters
+    ----------
+    est : BaseEstimator
+        A fitted classification or regression model.
+    target_variables : array-like, dtype=int
+        The target features for which the partial dependency should be
+        computed (size should be smaller than 3 for visual renderings).
+    grid : array-like, shape=(n_points, len(target_variables))
+        The grid of ``target_variables`` values for which the
+        partial dependency should be evaluated (either ``grid`` or ``X``
+        must be specified).
+    X : array-like, shape=(n_samples, n_features)
+        The data on which ``est`` was trained.
+    output : int, optional (default=None)
+        The output index to use for multi-output estimators.
+
+    Returns
+    -------
+    pdp : array, shape=(n_classes, n_points)
+        The partial dependence function evaluated on the ``grid``.
+        For regression and binary classification ``n_classes==1``.
+    """
+    n_samples = X.shape[0]
+    pdp = []
+    for row in range(grid.shape[0]):
+        X_eval = X.copy()
+        for i, variable in enumerate(target_variables):
+            X_eval[:, variable] = np.repeat(grid[row, i], n_samples)
+        if est._estimator_type == 'regressor':
```

Comment:
Shall look into it. Guess I felt that since the ``recursive`` method was thrown over to an outside module it made sense to give each one a method of its own. Can investigate for sure.